plutus-core/plutus-ir/src/PlutusIR/Transform/CaseOfCase.hs

```diff
@@ -0,0 +1,198 @@
+    underBindersMaybe (TermParam:arity) g (LamAbs a n ty t) = LamAbs a n ty <$> underBinders arity g t
+    underBindersMaybe (TypeParam:arity) g (TyAbs a ty k t)  = TyAbs a ty k <$> underBinders arity g t
+    underBindersMaybe _ _ _                                 = Nothing
+
+-- | Split a match, either a normal datatype match or a builtin 'match'.
+splitMatch :: (Ord fun, PLC.HasUnique name PLC.TermUnique, PLC.HasUnique tyname PLC.TypeUnique) =>
+  BuiltinsInfo uni fun
+  -> VarsInfo tyname name
+  -> Term tyname name uni fun a
+  -> Maybe (SplitMatchContext tyname name uni fun a, [Arity], SplitMatchContext tyname name uni fun a -> Term tyname name uni fun a)
+splitMatch binfo vinfo t = do
+  let (hd, args) = splitApplication t
+  (sm, arities, reconstruct) <- case hd of
+    (Var _ matcherName)     -> splitNormalDatatypeMatch vinfo matcherName args
+    (Builtin _ matcherName) -> splitBuiltinDatatypeMatch binfo matcherName args
+    _                       -> Nothing
+  -- wrangle the reconstruction function so it also puts the head back in, for
+  -- convenience of use elsewhere
+  pure (sm, arities, \sm' -> fillAppContext hd (reconstruct sm'))
+
+{- Note [Case-of-case and duplicating code]
+Case-of-case can duplicate code.
+
+Consider this schematic example:
+
+case (case s of { C1 a -> x; C2 b -> y; }) of
+  D1 -> w
+  D2 -> z
+
+-->
+
+case s of
+  C1 a -> case x of { D1 -> w; D2 -> z; }
+  C2 b -> case y of { D1 -> w; D2 -> z; }
+
+We duplicate the outer case (the matcher application, importantly including
+the branches) for every branch of the inner case.
+
+Any time we duplicate code we run the risk of exponential program growth,
+since we might apply case-of-case inside the duplicated parts of the
+program, thus leading to multiplicative growth. In particular, since we apply
+case-of-case in a bottom-up fashion, we can apply it inside a subterm (creating
+duplication), and then in the main term (multiplying the duplication) in a single
+pass. This is not _too_ bad so long as case-of-known-constructor will clear
+things up afterwards, but is something to be aware of.
+
+If the inner branch bodies are conapps (see Note [Case-of-case and conapps]) then
+that will reduce but not necessarily eliminate the duplication. Consider
+
+case (case s of { C1 a -> True; C2 b -> True; C3 c -> False }) of
+  True -> v
+  False -> w
+
+--> (case-of-case)
+
+case s of
+  C1 a -> case True of { True -> v; False -> w; }
+  C2 b -> case True of { True -> v; False -> w; }
+  C3 c -> case False of { True -> v; False -> w; }
+
+--> (case-of-known-constructor)
+
+case s of
+  C1 a -> v
+  C2 b -> v
+  C3 c -> w
+
+i.e. we still duplicate v!
+
+It's not all that uncommon to end up with multiple case branches that have the same
+result, e.g. if you write a wildcard fallthrough case in Haskell.
+-}
+
+{- Note [Case-of-case and conapps]
+Case-of-case will be much better if the bodies of the branches are all constructor applications.
```

Comment:
It's not jus t"will be much better", but to me it seems caseofcase is only applied if all the branches are conapps, correct?